File: app/src/main/java/com/jhughes/todoapp/data/domain/repo/TaskRepository.java
package com.jhughes.todoapp.data.domain.repo;

import com.jhughes.todoapp.data.domain.model.Task;
import com.jhughes.todoapp.data.local.repo.TaskDataSource;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class TaskRepository {

    private final TaskDataSource localDataSource;

    private Map<String, Task> tasksMap = new LinkedHashMap<>();;

    public TaskRepository(TaskDataSource taskDataSource) {
        this.localDataSource = taskDataSource;

        localDataSource.getTasks(new GetTasksCallback() {
            @Override
            public void onComplete(List<Task> tasks) {
                refreshCache(tasks);
            }
        });
    }

    public void getTasks(final GetTasksCallback callback) {

        if (tasksMap != null) {
            callback.onComplete(new ArrayList<>(tasksMap.values()));
        } else {
            localDataSource.getTasks(new GetTasksCallback() {
                @Override
                public void onComplete(List<Task> tasks) {
                    refreshCache(tasks);
                }
            });
        }
    }

    public void getTask(int taskId, final GetTaskCallback callback) {

        if (tasksMap != null) {
            final String id = String.valueOf(taskId);

            if (tasksMap.containsKey(id)) {
                callback.onComplete(tasksMap.get(id));
            }
        } else {
            localDataSource.getTask(taskId, new GetTaskCallback() {
                @Override
                public void onComplete(Task task) {
                    callback.onComplete(task);
                }
            });
        }
    }

    public void addTask(String description, final GetTaskCallback callback) {
        localDataSource.addTask(description, new GetTaskCallback() {
            @Override
            public void onComplete(Task task) {
                tasksMap.put(String.valueOf(task.getId()), task);
                callback.onComplete(task);
            }
        });
    }

    public void completeTask(String taskId) {
        if (tasksMap.containsKey(taskId)) {
            Task task = tasksMap.get(taskId);
            task.setComplete(true);
            tasksMap.put(taskId, task);

            localDataSource.completeTask(task);
        }
    }

    public void activateTask(String taskId) {
        if (tasksMap.containsKey(taskId)) {
            Task task = tasksMap.get(taskId);
            task.setComplete(false);
            tasksMap.put(taskId, task);

            localDataSource.activateTask(task);
        }
    }

    public void clearTasks() {
        tasksMap.clear();
        localDataSource.clearTasks();
    }

    private void refreshCache(List<Task> tasks) {
        tasksMap.clear();

        for (Task task : tasks) {
            tasksMap.put(String.valueOf(task.getId()), task);
        }
    }

    public interface GetTasksCallback {
        void onComplete(List<Task> tasks);
    }

    public interface GetTaskCallback {
        void onComplete(Task task);
    }
}

File: app/src/main/java/com/jhughes/todoapp/data/domain/repo/TaskRepository.kt
package com.jhughes.todoapp.data.domain.repo

import com.jhughes.todoapp.data.domain.model.Task
import com.jhughes.todoapp.data.local.repo.TaskDataSource
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TaskRepository @Inject constructor(private val localDataSource: TaskDataSource) {

    private var cachedTasks : List<Task>? = null

    fun getTasks(callback: (List<Task>) -> Unit) {
        with(cachedTasks) {
            if(this != null) {
                callback(this)
            } else {
                localDataSource.getTasks { tasks ->
                    cachedTasks = tasks
                    callback(tasks)
                }
            }
        }
    }

    fun getTask(taskId: Int, callback: (Task?) -> Unit) {
        with(cachedTasks) {
            if (this != null) {
                val task = find { it.id == taskId }
                callback(task)
            } else {
                getTasks { tasks ->
                    val task = tasks.find { it.id == taskId }
                    callback(task)
                }
            }
        }
    }

    fun addTask(description: String, callback: (Task) -> Unit) {
        localDataSource.addTask(description) { task ->
            cachedTasks = cachedTasks?.toMutableList()?.apply {
                add(task)
            }
            callback(task)
        }
    }

    fun completeTask(taskId: Int) {
        cachedTasks?.find { it.id == taskId }?.let {
            it.isComplete = true
            localDataSource.completeTask(it)
        }
    }

    fun activateTask(taskId: Int) {
        cachedTasks?.find { it.id == taskId }?.let {
            it.isComplete = false
            localDataSource.activateTask(it)
        }
    }

    fun clearTasks() {
        cachedTasks = null
        localDataSource.clearTasks()
    }
}
